src/DotVVM.Framework/ViewModel/Serialization/DotvvmDictionaryConverter.cs

```diff
@@ -0,0 +1,80 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using DotVVM.Framework.Utils;
+using Newtonsoft.Json;
+
+namespace DotVVM.Framework.ViewModel.Serialization
+{
+    public class DotvvmDictionaryConverter : JsonConverter
+    {
+        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+        {
+            if (value == null)
+            {
+                writer.WriteNull();
+            }
+            else
+            {
+                var dict = value as IDictionary;
+
+                var attrs = value.GetType().GetGenericArguments();
+                var keyValuePair = typeof(KeyValuePair<,>).MakeGenericType(attrs);
+                var listType = typeof(List<>).MakeGenericType(keyValuePair);
+
+
+                var keyEnumerator = dict.Keys.GetEnumerator();
+                var valuesEnumerator = dict.Values.GetEnumerator();
+
+
+                var list = Activator.CreateInstance(listType);
+                var invokeMethod = listType.GetMethod(nameof(List<object>.Add));
+                while (keyEnumerator.MoveNext() && valuesEnumerator.MoveNext())
+                {
+                    var item = Activator.CreateInstance(keyValuePair, keyEnumerator.Current, valuesEnumerator.Current);
+                    invokeMethod.Invoke(list, new[] { item });
+                }
```

Comment:
I think this could be simplified using `dynamic` (not checked it works, though):

```suggestion
                KeyValuePair<T, U>[] toArray<T, U>(IEnumerable<KeyValuePair<T, U>> x) =>
                       x.ToArray()
                object list = toArray((dynamic) value);
```


It will resolve the generic parameters by itself.

Reply:
I have tried and unfortunately does not work.